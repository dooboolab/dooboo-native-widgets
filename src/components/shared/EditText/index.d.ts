import * as React from 'react';
import { ViewStyle, ImageSourcePropType } from 'react-native';

interface Props {
  parentTestID?: string;
  testID?: string;
  errorTestID?: string;
  style?: ViewStyle;
  label?: string;
  textStyle?: TextStyle;
  errorText?: string;
  text?: string;
  placeholder?: string;
  placeholderTextColor?: string;
  secureTextEntry?: boolean;
  onSubmitEditing?: (e: NativeSyntheticEvent<any>) => void;
  onTextChanged?: (text: string) => void;
}

const EditText: React.SFC<Props> = (props) => {

};

export default EditText;
