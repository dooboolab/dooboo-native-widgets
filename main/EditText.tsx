import {Platform, StyleSheet, Text, TextInput, View} from 'react-native';
import React, {FC, useState} from 'react';
import type {StyleProp, TextInputProps, ViewStyle} from 'react-native';

const defaultStyle: ViewStyle = {
  alignSelf: 'stretch',
  flex: 1,

  flexDirection: 'column',
};

const defaultRowStyles = StyleSheet.create({
  container: {
    alignSelf: 'stretch',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    borderBottomWidth: 0.3,
    borderBottomColor: '#666666',
  },
  labelText: {
    fontSize: 14,
    color: '#b9b9c4',
    width: 100,
  },
  input: {
    paddingVertical: 12,
    paddingHorizontal: 8,
    fontSize: 14,
    fontWeight: 'bold',
    flex: 1,
    color: '#2c374e',
  },
  errorText: {
    marginTop: 4,
    fontSize: 12,
    color: '#991111',
  },
});

const defaultColumnStyles = StyleSheet.create({
  container: {
    alignSelf: 'stretch',
    justifyContent: 'space-between',
    borderBottomWidth: 0.3,
    borderBottomColor: '#666666',

    flexDirection: 'column',
  },
  labelText: {
    fontSize: 14,
    color: '#b9b9c4',
  },
  input: {
    paddingVertical: 12,
    fontSize: 14,
    fontWeight: 'bold',
    flex: 1,
    color: '#2c374e',
  },
  errorText: {
    marginTop: 8,
    fontSize: 12,
    color: '#991111',
  },
});

type StylesType = Partial<StyleSheet.NamedStyles<typeof defaultRowStyles>>;

type Props = {
  testID?: TextInputProps['testID'];
  textInputProps?: TextInputProps;
  style?: StyleProp<ViewStyle>;
  styles?: StylesType;
  labelText?: string;
  errorText?: string;
  value?: TextInputProps['value'];
  onChange?: TextInputProps['onChange'];
  onChangeText?: TextInputProps['onChangeText'];
  placeholder?: TextInputProps['placeholder'];
  placeholderTextColor?: TextInputProps['placeholderTextColor'];
  onFocus?: TextInputProps['onFocus'];
  onBlur?: TextInputProps['onBlur'];
  editable?: TextInputProps['editable'];
  autoCapitalize?: TextInputProps['autoCapitalize'];
  secureTextEntry?: TextInputProps['secureTextEntry'];
  onSubmitEditing?: TextInputProps['onSubmitEditing'];
  focusColor?: string;
  errorColor?: string;
  disableColor?: string;
  type?: 'row' | 'column';
};

const EditText: FC<Props> = ({
  testID,
  textInputProps,
  style,
  styles,
  labelText = '',
  errorText = '',
  value = '',
  placeholder,
  placeholderTextColor,
  onChange,
  onChangeText,
  onFocus,
  onBlur,
  autoCapitalize = 'none',
  secureTextEntry = false,
  editable = true,
  focusColor = '#9A77FF',
  errorColor = '#ff7676',
  disableColor = '#999999',
  type = 'column',
}) => {
  const [focused, setFocused] = useState(false);

  const defaultStyles = type === 'row' ? defaultRowStyles : defaultColumnStyles;

  return (
    <View style={[defaultStyle, style]}>
      <View
        style={[
          defaultStyles.container,
          {
            borderColor: errorText
              ? errorColor
              : focused
              ? focusColor
              : disableColor,
            borderBottomColor: errorText
              ? errorColor
              : focused
              ? focusColor
              : disableColor,
          },
        ]}>
        {/* {labelText ? ( */}
        <Text
          style={[
            defaultStyles.labelText,
            styles?.labelText,
            {
              color: errorText
                ? errorColor
                : focused
                ? focusColor
                : disableColor,
            },
          ]}>
          {labelText}
        </Text>
        {/* ) : null} */}
        <TextInput
          {...textInputProps}
          testID={testID}
          autoCapitalize={autoCapitalize}
          secureTextEntry={secureTextEntry}
          style={[
            defaultStyles.input,
            styles?.input,
            // @ts-ignore
            Platform.OS === 'web' && {outlineWidth: 0},
          ]}
          editable={editable}
          onFocus={(e) => {
            setFocused(true);
            onFocus?.(e);
          }}
          onBlur={(e) => {
            setFocused(false);
            onBlur?.(e);
          }}
          value={value}
          placeholder={placeholder}
          placeholderTextColor={placeholderTextColor}
          onChange={onChange}
          onChangeText={onChangeText}
        />
      </View>
      {errorText ? (
        <Text
          style={[
            defaultStyles.errorText,
            styles?.errorText,
            {color: errorColor},
          ]}>
          {errorText}
        </Text>
      ) : null}
    </View>
  );
};

export {EditText};
