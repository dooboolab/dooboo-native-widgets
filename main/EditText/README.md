# EditText

> EditText component is an enhanced version on pure react-native [TextInput] component.</br>

![image](https://user-images.githubusercontent.com/58724686/87130683-4121b100-c2ce-11ea-8663-3f6e7d380a6d.gif)

## Props

|                      | necessary | types                  | default              |
| -------------------- | --------- | ---------------------- | -------------------- |
| labelPosition        |           | 'row'                  |                      |
| placeholder          |           | string                 |         `text`       |
| placeholderTextColor |           | object                 |                      |
| containerStyle       |           | `ViewStyle`            |                      |
| textInputStyle       |           | `ViewStyle`            |                      |
| focusColor           |           | string                 |                      |
| labelText            |           | string                 |                      |
| labelTextStyle       |           | `TextStyle`            |                      |
| value                |           | `TextInputProps`       |                      |
| isErrored            |           | boolean                |        `false`       |
| errorMessage         |           | `React.ReactElement`   |                      |
| errorStyle           |           | `ViewStyle`            |                      |
| numberOfLine         |           |  number                |                      |
| secureTextEntry      |           |  boolean               |                      |
| onFocus              |           |  ()=> {}               |                      |
| onBlur               |           |  ()=> {}               |                      |
| onChangeText         |           |  ()=> {}               |                      |
| onSubmitEditing      |           |  ()=> {}               |                      |

## Installation

```sh
yarn add dooboo-ui
```

## Description
- When you want to use Row type, you should set labelPosition props to 'row'. 
  You can remove the border with containerStyle porps in row type. but In default, you can remove the border with textInputStyle props. 

## Usage
  This is use case for signIn view.

```javascript

import {EditText} from 'dooboo-ui';

const UseCase = ():React.ReactElement => {
  const [email, setEmail] = useState<string>('');
  const [password, setPassword] = useState<string>('');
  const [isEmailErrored, setEmailError] = useState<boolean>(false);
  const [isPswErrored, setPswError] = useState<boolean>(false);

  const onSignIn = (): void => {
    if (email === '' && password === '') {
      setEmailError(true);
      setPswError(true);
    } else if (email === '') {
      setEmailError(true);
    } else if (password === '') {
      setPswError(true);
    }
  };

  return (
    <SafeAreaView>
      <ScrollView
        contentContainerStyle={{
          marginTop: 8,
          alignSelf: 'stretch',
          paddingHorizontal: 20,
          paddingVertical: 100,
        }}
      >
        <View
          style={{
            display: 'flex',
            flexDirection: 'column',
            alignItems: 'center',
            padding: 20,
            paddingTop: 10,
          }}
        >
          <Text
            style={{
              fontWeight: 'bold',
              fontSize: 36,
              lineHeight: 53,
              color: '#1B1B21',
            }}
          >
            SignUp
          </Text>
        </View>
        <View
          style={{
            justifyContent: 'center',
            alignItems: 'center',
          }}
        >
          <EditText
            labelText={'Email'}
            value={email}
            isErrored={isEmailErrored}
            errorStyle={{ borderColor: '#E54E4E' }}
            errorMessage={
              <Text style={{ color: '#E54E4E', marginTop: 8 }}>
                Invaild Email address
              </Text>
            }
            placeholder={'dooboolab@gmail.com'}
            placeholderTextColor={'#707683'}
            onChangeText={(email): void => { setEmail(email); }}
            containerStyle={{
              marginBottom: 30,
              width: 528,
            }}
            labelTextStyle={{
              color: '#323C47',
            }}
            onFocus={(): void => setEmailError(false)}
          />
          <EditText
            labelText={'Password'}
            value={password}
            isErrored={isPswErrored}
            errorStyle={{ borderColor: '#E54E4E' }}
            errorMessage={
              <Text style={{ color: '#E54E4E', marginTop: 8 }}>
                Invaild Password
              </Text>
            }
            placeholder={'********'}
            placeholderTextColor={'#707683'}
            onChangeText={(password): void => { setPassword(password); }}
            secureTextEntry={true}
            containerStyle={{
              width: 528,
            }}
            labelTextStyle={{
              color: '#323C47',
            }}
            onFocus={(): void => setPswError(false)}
          />
        </View>
        <View
          style={{
            alignItems: 'center',
            marginTop: 25,
          }}
        >
          <Text
            style={{
              fontSize: 14,
              lineHeight: 18,
              color: '#231F20',
            }}
          >
            Agree to terms and conditions
          </Text>
        </View>
        <Button
          style={{
            borderWidth: 1,
            borderColor: '#609FFF',
            borderRadius: 6,
            marginTop: 40,
            width: 528,
          }}
          text="이메일 인증하기"
          textStyle={{
            color: '#609FFF',
            fontWeight: 'bold',
            fontSize: 10,
            lineHeight: 15,
            letterSpacing: -0.1,
          }}
          onPress={(): void => onSignIn()}
        />
      </ScrollView>
    </SafeAreaView>
  );
};
```
