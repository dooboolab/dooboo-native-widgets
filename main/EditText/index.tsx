import {
  KeyboardTypeOptions,
  NativeSyntheticEvent,
  Platform,
  TextInputProps,
  TextInputSubmitEditingEventData,
  TextStyle,
  ViewStyle,
} from 'react-native';
import React, { FC, ReactElement, useState } from 'react';

import styled from 'styled-components/native';

const StyledRowContent = styled.View`
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  width: 100%;
`;

const StyledRowLabel = styled.Text`
  font-size: 14px;
  font-weight: bold;
  color: #b9b9c4;
`;

const StyledRowInput = styled.TextInput`
  padding-top: 16px;
  padding-bottom: 16px;
  font-size: 14px;
  font-weight: bold;
  flex: 1;
  height: 100%;
  color: #2c374e;
  ${Platform.OS === 'web' && { 'outline-style': 'none' }}
`;

const Container = styled.View`
  flex-direction: column;
  align-self: stretch;
`;

const StyledLine = styled.View`
  flex-direction: row;
  align-items: center;
  width: 100%;
`;

const StyledLabel = styled.Text`
  margin-bottom: 5px;
  font-size: 14px;
  font-weight: 500;
  color: #b9b9c4;
`;

const StyledIcon = styled.View`
  align-items: center;
  justify-content: center;
`;

const StyledTextInput = styled.TextInput`
  padding-top: 15px;
  padding-bottom: 15px;
  font-size: 15px;
  font-weight: 500;
  flex: 1;
  ${Platform.OS === 'web' && { 'outline-style': 'none' }}
`;

const StyledInvalidText = styled.Text`
  margin: 0px 2px;
  font-size: 12px;
  font-weight: 500;
  margin-top: 5px;
  color: #ff8989;
`;

interface Props {
  testID?: string;
  errorTestID?: string;
  type?: string;
  isRow?: boolean;
  style?: ViewStyle;
  label?: string;
  labelTextStyle?: TextStyle;
  labelWidth?: number;
  value?: TextInputProps['value'];
  contentStyle?: ViewStyle;
  inputContainerType?: string;
  inputContainerRadius?: number;
  borderStyle?: ViewStyle;
  borderWidth?: number;
  borderColor?: string;
  textStyle?: TextStyle;
  placeholder?: TextInputProps['placeholder'];
  placeholderTextColor?: TextInputProps['placeholderTextColor'];
  secureTextEntry?: TextInputProps['secureTextEntry'];
  onChangeText?: TextInputProps['onChangeText'];
  onSubmitEditing?: (
    e: NativeSyntheticEvent<TextInputSubmitEditingEventData>,
  ) => void;
  leftElement?: ReactElement;
  leftElementStyle?: ViewStyle;
  rightElement?: ReactElement;
  rightElementStyle?: ViewStyle;
  focusedLabelStyle?: TextStyle;
  focusedBorderWidth?: number;
  focusColor?: string;
  errorColor?: string;
  autoCapitalize?: TextInputProps['autoCapitalize'];
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  textInputProps?: TextInputProps | any;
  onFocus?: () => void;
  onBlur?: () => void;
  errorText?: string;
  errorTextStyle?: TextStyle;
  keyboardType?: KeyboardTypeOptions;
  numberOfLines?: number;
  multiline?: boolean;
}

export enum EditTextInputType {
  DEFAULT = 'default',
  ROW = 'row',
  BOX = 'box',
  ROW_BOX = 'rowBox',
}

const EditText: FC<Props> = (props) => {
  const [focused, setFocus] = useState(false);

  const {
    testID,
    errorTestID,
    type = EditTextInputType.DEFAULT,
    style,
    label,
    labelTextStyle,
    labelWidth = 110,
    value,
    contentStyle,
    borderStyle,
    borderWidth = 0.6,
    borderColor = '#eaeaf9',
    textStyle = Platform.select({
      ios: {
        paddingVertical: 16,
      },
      android: {
        paddingVertical: 10,
      },
    }),
    placeholder,
    placeholderTextColor,
    secureTextEntry,
    onChangeText,
    onSubmitEditing,
    rightElement,
    rightElementStyle,
    focusedLabelStyle = { fontWeight: 'bold' },
    focusedBorderWidth = 1,
    focusColor = '#79B3F5',
    errorColor = '#FF8989',
    autoCapitalize = 'none',
    textInputProps,
    onFocus,
    onBlur,
    errorText,
    errorTextStyle,
    keyboardType,
    numberOfLines,
    multiline = false,
  } = props;

  switch (type) {
    case EditTextInputType.DEFAULT:
    default:
      return (
        <Container style={style}>
          <StyledLabel
            style={[
              labelTextStyle,
              errorText
                ? { color: errorColor }
                : focused && [
                  { color: focusColor },
                  focusedLabelStyle,
                ],
            ]}
          >
            {label}
          </StyledLabel>
          <StyledTextInput
            {...textInputProps}
            testID={testID}
            autoCapitalize={autoCapitalize}
            onFocus={(): void => {
              setFocus(true);
              onFocus && onFocus();
            }}
            onBlur={(): void => {
              setFocus(false);
              onBlur && onBlur();
            }}
            placeholder={placeholder}
            placeholderTextColor={placeholderTextColor}
            value={value}
            style={textStyle}
            numberOfLines={numberOfLines}
            onChangeText={onChangeText}
            secureTextEntry={secureTextEntry}
            onSubmitEditing={onSubmitEditing}
            multiline={multiline}
          />
          {
            borderWidth && <StyledLine
              style={[
                borderStyle,
                { borderBottomWidth: borderWidth, borderColor: borderColor },
                errorText
                  ? { borderColor: errorColor }
                  : focused && [
                    { borderColor: focusColor },
                    { borderBottomWidth: focusedBorderWidth },
                  ],
              ]}
            />
          }
          {errorText ? (
            <StyledInvalidText
              testID={errorTestID}
              style={[{ color: errorColor }, errorTextStyle]}>
              {`${errorText}`}
            </StyledInvalidText>
          ) : null}
        </Container>
      );
    case EditTextInputType.ROW:
      return (
        <Container style={style}>
          <StyledRowContent
            style={[
              { borderColor: borderColor, borderBottomWidth: borderWidth },
              borderStyle,
              errorText
                ? {
                  borderColor: errorColor,
                  borderBottomWidth: focusedBorderWidth,
                }
                : focused && {
                  borderColor: focusColor,
                  borderBottomWidth: focusedBorderWidth,
                },
              contentStyle,
            ]}
          >
            {label ? (
              <StyledRowLabel
                style={[
                  labelTextStyle,
                  errorText
                    ? [{ color: errorColor }, focusedLabelStyle]
                    : focused && [
                      { color: focusColor },
                      focusedLabelStyle,
                    ],
                  { width: labelWidth },
                ]}
              >
                {label}
              </StyledRowLabel>
            ) : null}
            <StyledRowInput
              {...textInputProps}
              testID={testID}
              style={[textStyle, { textAlign: 'left' }]}
              autoCapitalize={autoCapitalize}
              onFocus={(): void => {
                setFocus(true);
                if (onFocus) {
                  onFocus();
                }
              }}
              onBlur={(): void => {
                setFocus(false);
                if (onBlur) {
                  onBlur();
                }
              }}
              onSubmitEditing={onSubmitEditing}
              placeholder={placeholder}
              placeholderTextColor={placeholderTextColor}
              value={value}
              numberOfLines={numberOfLines}
              onChangeText={onChangeText}
              secureTextEntry={secureTextEntry}
              keyboardType={keyboardType}
              multiline={multiline}
            />
          </StyledRowContent>
          {errorText ? (
            <StyledInvalidText testID={errorTestID} style={errorTextStyle}>
              {errorText}
            </StyledInvalidText>
          ) : null}
        </Container>
      );

    case EditTextInputType.BOX:
      return (
        <Container style={style}>
          <StyledLabel
            style={[
              labelTextStyle,
              errorText
                ? { color: errorColor }
                : focused && [
                  { color: focusColor },
                  focusedLabelStyle,
                ],
            ]}
          >
            {label}
          </StyledLabel>
          <StyledRowContent
            style={[
              {
                borderWidth: borderWidth,
                borderColor: borderColor,
              },
              borderStyle,
              errorText
                ? { borderColor: errorColor, borderWidth: focusedBorderWidth }
                : focused && {
                  borderColor: focusColor,
                  borderWidth: focusedBorderWidth,
                },
              contentStyle,
            ]}
          >
            <StyledTextInput
              {...textInputProps}
              testID={testID}
              autoCapitalize={autoCapitalize}
              onFocus={(): void => {
                setFocus(true);
                if (onFocus) {
                  onFocus();
                }
              }}
              onBlur={(): void => {
                setFocus(false);
                if (onBlur) {
                  onBlur();
                }
              }}
              placeholder={placeholder}
              placeholderTextColor={placeholderTextColor}
              value={value}
              style={textStyle}
              onChangeText={onChangeText}
              secureTextEntry={secureTextEntry}
              onSubmitEditing={onSubmitEditing}
              multiline={multiline}
            />
            {rightElement && (
              <StyledIcon style={[{ width: 40 }, rightElementStyle]}>
                {rightElement}
              </StyledIcon>
            )}

          </StyledRowContent>
          {errorText ? (
            <StyledInvalidText
              testID={errorTestID}
              style={[
                {
                  color: errorColor,
                },
                errorTextStyle,
              ]}
            >
              {`${errorText}`}
            </StyledInvalidText>
          ) : null}
        </Container>
      );

    case EditTextInputType.ROW_BOX:
      return (
        <Container style={style}>
          <StyledRowContent
            style={[
              {
                borderWidth: borderWidth,
                borderColor: borderColor,
              },
              borderStyle,
              errorText
                ? { borderColor: errorColor, borderWidth: focusedBorderWidth }
                : focused && {
                  borderColor: focusColor,
                  borderWidth: focusedBorderWidth,
                },
            ]}
          >
            {label ? (
              <StyledRowLabel
                style={[
                  labelTextStyle,
                  errorText
                    ? [{ color: errorColor }, focusedLabelStyle]
                    : focused
                      ? [{ color: focusColor }, focusedLabelStyle]
                      : null,
                  { marginLeft: 15, width: labelWidth },
                ]}>
                {label}
              </StyledRowLabel>
            ) : null}
            <StyledRowInput
              {...textInputProps}
              testID={testID}
              style={[{ paddingRight: 15 }, textStyle]}
              autoCapitalize={autoCapitalize}
              onFocus={(): void => {
                setFocus(true);
                if (onFocus) {
                  onFocus();
                }
              }}
              onBlur={(): void => {
                setFocus(false);
                if (onBlur) {
                  onBlur();
                }
              }}
              onSubmitEditing={onSubmitEditing}
              placeholder={placeholder}
              placeholderTextColor={placeholderTextColor}
              value={value}
              numberOfLines={numberOfLines}
              onChangeText={onChangeText}
              secureTextEntry={secureTextEntry}
              keyboardType={keyboardType}
              multiline={multiline}
            />
          </StyledRowContent>
          {errorText ? (
            <StyledInvalidText testID={errorTestID} style={errorTextStyle}>
              {errorText}
            </StyledInvalidText>
          ) : null}
        </Container>
      );
  }
};

export { EditText };
