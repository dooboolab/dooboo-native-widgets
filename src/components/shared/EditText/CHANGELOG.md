[0.4.4]
- Support boxes [#177](https://github.com/dooboolab/dooboo-ui-native/pull/117).

[0.4.1]
- Support row aligned.

[0.3.0]
- First release.
