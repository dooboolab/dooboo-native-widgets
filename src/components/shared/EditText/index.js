import React, { useState } from 'react';
import styled from 'styled-components/native';
import PropTypes from 'prop-types';

import {
  StyleSheet,
  Text,
  View,
  TextInput,
  TextStyle,
  ViewPropTypes,
  ViewStyle,
  Platform,
  NativeSyntheticEvent,
} from 'react-native';

const styles = StyleSheet.create({
  container: {
    flexDirection: 'column',
    alignSelf: 'stretch',
    width: '100%',
  },
  underline: {
    borderWidth: 0.6,
    borderColor: '#eaeaf9',
  },
  label: {
    fontSize: 14,
    fontWeight: '500',
    color: '#b9b9c4',
  },
  input: Platform.select({
    ios: {
      paddingTop: 15,
      paddingBottom: 10,
      fontSize: 15,
      fontWeight: '500',
      color: '#2C374E',
    },
    android: {
      paddingLeft: 0,
      paddingBottom: 5,
      fontSize: 15,
      fontWeight: '500',
      color: '#2C374E',
    },
  }),
  errorText: {
    fontSize: 12,
    fontWeight: '500',
    marginTop: 5,
    color: '#ff8989',
  },
});

function EditText(props) {
  const [focused, setFocus] = useState(false);

  return (
    <View
      testID={ props.parentTestID }
      style={[
        styles.container,
        props.style,
      ]}
    >
      {
        props.label
          ? <Text style={[
            styles.label,
            props.errorText
              ? { color: '#FF8989' }
              : focused
                ? { color: '#79B3F5' }
                : null,
          ]}
          >
            { props.label }
          </Text>
          : null
      }
      <TextInput
        testID={props.testID}
        style={[
          styles.input,
          props.textStyle,
        ]}
        autoCapitalize={'none'}
        onFocus={ () => setFocus(true) }
        onBlur={ () => setFocus(false) }
        onSubmitEditing={ props.onSubmitEditing }
        placeholder={ props.placeholder }
        placeholderTextColor={ props.placeholderTextColor }
        value={ props.text }
        onChangeText={ props.onTextChanged}
        secureTextEntry={ props.secureTextEntry }
      >
      </TextInput>
      <View style={[
        styles.underline,
        focused
          ? { borderColor: '#79B3F5' }
          : props.errorText
            ? { borderColor: '#FF8989' }
            : null,
      ]}
      />
      {
        props.errorText
          ? <Text
            testID={props.errorTestID}
            style={styles.errorText}
          >
            {props.errorText}
          </Text>
          : null
      }
    </View>
  );
}

EditText.propTypes = {
  parentTestID: PropTypes.string,
  testID: PropTypes.string,
  errorTestID: PropTypes.string,
  style: ViewPropTypes.style,
  label: PropTypes.string,
  textStyle: Text.propTypes.style,
  errorText: PropTypes.string,
  text: PropTypes.string,
  placeholder: PropTypes.string,
  placeholderTextColor: PropTypes.string,
  secureTextEntry: PropTypes.bool,
  onSubmitEditing: PropTypes.func,
  onTextChanged: PropTypes.func,
};

export default EditText;
