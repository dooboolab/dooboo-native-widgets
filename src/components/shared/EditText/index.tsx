import {
  NativeSyntheticEvent,
  Platform,
  StyleProp,
  StyleSheet,
  Text,
  TextInput,
  TextStyle,
  View,
  ViewStyle,
} from 'react-native';
import React, { useState } from 'react';

const styles = StyleSheet.create({
  container: {
    flexDirection: 'column',
    alignSelf: 'stretch',
    width: '100%',
  },
  underline: {
    borderWidth: 0.6,
    borderColor: '#eaeaf9',
  },
  label: {
    fontSize: 14,
    fontWeight: '500',
    color: '#b9b9c4',
  },
  input: Platform.select({
    ios: {
      paddingTop: 15,
      paddingBottom: 10,
      fontSize: 15,
      fontWeight: '500',
      color: '#2C374E',
    },
    android: {
      paddingLeft: 0,
      paddingBottom: 5,
      fontSize: 15,
      fontWeight: '500',
      color: '#2C374E',
    },
  }),
  errorText: {
    fontSize: 12,
    fontWeight: '500',
    marginTop: 5,
    color: '#ff8989',
  },
});

interface Props {
  parentTestID?: string;
  testID?: string;
  errorTestID?: string;
  style?: StyleProp<ViewStyle>;
  label?: string;
  textStyle?: StyleProp<TextStyle>;
  errorText?: string;
  text?: string;
  placeholder?: string;
  placeholderTextColor?: string;
  secureTextEntry?: boolean;
  onSubmitEditing?: (e: NativeSyntheticEvent<any>) => void;
  onTextChanged?: (text: string) => void;
}

function EditText(props: Props) {
  const [focused, setFocus] = useState(false);

  return (
    <View testID={props.parentTestID} style={[styles.container, props.style]}>
      {props.label ? (
        <Text
          style={[
            styles.label,
            props.errorText
              ? { color: '#FF8989' }
              : focused
                ? { color: '#79B3F5' }
                : null,
          ]}
        >
          {props.label}
        </Text>
      ) : null}
      <TextInput
        testID={props.testID}
        style={[styles.input, props.textStyle]}
        autoCapitalize={'none'}
        onFocus={() => setFocus(true)}
        onBlur={() => setFocus(false)}
        onSubmitEditing={props.onSubmitEditing}
        placeholder={props.placeholder}
        placeholderTextColor={props.placeholderTextColor}
        value={props.text}
        onChangeText={props.onTextChanged}
        secureTextEntry={props.secureTextEntry}
      ></TextInput>
      <View
        style={[
          styles.underline,
          focused
            ? { borderColor: '#79B3F5' }
            : props.errorText
              ? { borderColor: '#FF8989' }
              : null,
        ]}
      />
      {props.errorText ? (
        <Text testID={props.errorTestID} style={styles.errorText}>
          {props.errorText}
        </Text>
      ) : null}
    </View>
  );
}

export default EditText;
