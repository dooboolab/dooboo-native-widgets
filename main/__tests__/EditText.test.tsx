import React, { ReactElement } from 'react';
import {
  RenderResult,
  act,
  cleanup,
  fireEvent,
  render,
} from '@testing-library/react-native';
import EditText from '../EditText';

let props: any;
let component: ReactElement;
let testingLib: RenderResult;

const createTestProps = (
  obj?: Record<string, unknown>,
): Record<string, unknown> => ({
  ...obj,
});

describe('[EditText] render default component test', () => {
  it('should render without crasing', () => {
    component = <EditText />;
    testingLib = render(component);

    expect(testingLib.baseElement).toMatchSnapshot();
  });

  it('should render with error message when isErrored props is true', () => {
    props = createTestProps({
      isErrored: true,
    });

    component = <EditText {...props} />;
    testingLib = render(component);

    expect(testingLib.baseElement).toMatchSnapshot();
  });

  it('should render Row type component when labelPosition props is row', () => {
    props = createTestProps({
      labelPosition: 'row',
    });

    component = <EditText {...props} />;
    testingLib = render(component);
    expect(testingLib.baseElement).toMatchSnapshot();
  });
});

describe('[EditText] event test', () => {
  beforeAll(() => {
    props = createTestProps();
    testingLib = render(<EditText />);
  });

  it('should change borderColor when focused', () => {
    const textInput = testingLib.getByTestId('TextInput-test');

    act(() => {
      fireEvent.focus(textInput);
    });
    expect(textInput.props.style[2].borderColor).toEqual('#109CF1');
  });

  it('should change borderColor when blured', () => {
    const textInput = testingLib.getByTestId('TextInput-test');

    act(() => {
      fireEvent.blur(textInput);
    });

    expect(textInput.props.style[0].borderColor).toEqual('#e0e0e0');
  });

  afterAll((done) => {
    cleanup();
    done();
  });
});
