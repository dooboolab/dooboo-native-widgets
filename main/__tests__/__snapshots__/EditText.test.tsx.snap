// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[EditText] render default component test should render Row type component when labelPosition props is row 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderColor": "#e0e0e0",
            "borderRadius": 6,
            "borderWidth": 1,
            "flexDirection": "row",
            "paddingLeft": 20,
            "width": 335,
          },
          undefined,
          false,
        ]
      }
      testID="EditText-test"
    >
      <Text
        style={
          Array [
            Object {
              "color": "#828282",
              "fontSize": 14,
              "fontStyle": "normal",
              "fontWeight": "800",
              "lineHeight": 19,
            },
          ]
        }
      >
        Label
      </Text>
      <TextInput
        allowFontScaling={true}
        placeholder="text"
        placeholderTextColor="#BDBDBD"
        rejectResponderTermination={true}
        secureTextEntry={false}
        style={
          Array [
            Object {
              "color": "#4f4f4f",
              "fontSize": 20,
              "fontStyle": "normal",
              "fontWeight": "500",
              "height": 56,
              "lineHeight": 27,
              "paddingBottom": 15,
              "paddingLeft": 20,
              "paddingRight": 20,
              "paddingTop": 14,
              "width": "100%",
            },
          ]
        }
        testID="TextInput-row-test"
        underlineColorAndroid="transparent"
      />
    </View>
  </View>
</View>
`;

exports[`[EditText] render default component test should render with error message when isErrored props is true 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "flexDirection": "column",
            "width": 335,
          },
        ]
      }
      testID="EditText-test"
    >
      <Text
        style={
          Array [
            Object {
              "color": "#828282",
              "fontSize": 14,
              "fontStyle": "normal",
              "fontWeight": "800",
              "lineHeight": 19,
              "marginBottom": 6,
            },
          ]
        }
      >
        Label
      </Text>
      <TextInput
        allowFontScaling={true}
        placeholder="text"
        placeholderTextColor="#BDBDBD"
        rejectResponderTermination={true}
        secureTextEntry={false}
        style={
          Array [
            Object {
              "borderColor": "#e0e0e0",
              "borderRadius": 6,
              "borderWidth": 1,
              "color": "#4f4f4f",
              "fontSize": 20,
              "fontStyle": "normal",
              "fontWeight": "500",
              "height": 56,
              "lineHeight": 27,
              "paddingBottom": 15,
              "paddingLeft": 20,
              "paddingRight": 20,
              "paddingTop": 14,
              "width": "100%",
            },
            undefined,
            Object {
              "borderColor": "#E54E4E",
            },
          ]
        }
        testID="TextInput-test"
        underlineColorAndroid="transparent"
      />
      <Text
        style={
          Object {
            "color": "#E54E4E",
            "marginTop": 8,
          }
        }
      >
        Default error message
      </Text>
    </View>
  </View>
</View>
`;

exports[`[EditText] render default component test should render without crasing 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "flexDirection": "column",
            "width": 335,
          },
        ]
      }
      testID="EditText-test"
    >
      <Text
        style={
          Array [
            Object {
              "color": "#828282",
              "fontSize": 14,
              "fontStyle": "normal",
              "fontWeight": "800",
              "lineHeight": 19,
              "marginBottom": 6,
            },
          ]
        }
      >
        Label
      </Text>
      <TextInput
        allowFontScaling={true}
        placeholder="text"
        placeholderTextColor="#BDBDBD"
        rejectResponderTermination={true}
        secureTextEntry={false}
        style={
          Array [
            Object {
              "borderColor": "#e0e0e0",
              "borderRadius": 6,
              "borderWidth": 1,
              "color": "#4f4f4f",
              "fontSize": 20,
              "fontStyle": "normal",
              "fontWeight": "500",
              "height": 56,
              "lineHeight": 27,
              "paddingBottom": 15,
              "paddingLeft": 20,
              "paddingRight": 20,
              "paddingTop": 14,
              "width": "100%",
            },
            undefined,
            false,
          ]
        }
        testID="TextInput-test"
        underlineColorAndroid="transparent"
      />
    </View>
  </View>
</View>
`;
